libd/src/d/semantic/expression.d

```diff
@@ -328,10 +341,76 @@ private:
 				assert(0, "Unorderd comparisons are not implemented.");
 		}
 	}
-	
+
 public:
 	Expression visit(AstBinaryExpression e) {
-		return buildBinary(e.location, e.op, visit(e.lhs), visit(e.rhs));
+		auto lhs = visit(e.lhs);
+		auto rhs = visit(e.rhs);
+		import d.context.name;
+
+		struct OverloadInfo {
+			Name name;
+			bool isTemplate;
+		}
+
+		const OverloadInfo overloadInfo(const AstBinaryOp op) pure {
+			switch (op) with (AstBinaryOp) {
+				case Assign :
+					return OverloadInfo(BuiltinName!"opAssign", false);
+				case Equal, NotEqual :
+					return OverloadInfo(BuiltinName!"opEquals", false);
+				case  Add, Sub,	Mul, Div, Mod, Pow :
+					return OverloadInfo(BuiltinName!"opBinary", true);
+				default : 
+					return OverloadInfo (BuiltinName!"", false);
+			}
+		}
+		
+		Expression handleOverloadedBinaryOp (Location location, AstBinaryOp op, Expression lhs, Expression rhs) {
+			Expression call;			
+
+			auto oInfo = overloadInfo(e.op);
+			if (op.isAssign) {
+				// oInfo = OverloadInfo(BuiltinName!"opOpAssign", true);
+				// if we cannot find opOpAssign let's call buildOpOpAssign with ourselfs as delegate
+				// return buildOpOpAssign(location, op, lhs, rhs, &handleOverloadedBinaryOp); 
+			}
+
+			auto resolvedOpSymbol = lhs.type.aggregate.resolve(e.location, oInfo.name);
+
+			if (auto os = cast(OverloadSet) resolvedOpSymbol) {
+				if (oInfo.isTemplate) {
+					// we maybe want to do diffrent things for templates in the future 
+					assert(0, "Templates like opBinary not supported right now");
+				} else {
+					call = callOverloadSet(e.location, os, [lhs, rhs]);
+				}
+			} else if (auto f = cast(Function) resolvedOpSymbol) {
+				call = handleCall(e.location, build!FunctionExpression(f.location, f), [lhs, rhs]);
+				assert(0, "I am surprised we got here... If we do just remove this assert");
+			}
+
+			if (e.op == AstBinaryOp.NotEqual && call) {
+				call = build!UnaryExpression(
+					e.location,
+					Type.get(BuiltinType.Bool),
+					UnaryOp.Not,
+					call,
+				);
+			}
+
+			return call;
+		}
+
+		auto oInfo = overloadInfo(e.op);
+		if (lhs.type.isAggregate() && oInfo.name != BuiltinName!"") { 
+			auto call =  handleOverloadedBinaryOp (e.location, e.op, lhs, rhs);
+			if (call) { 
+				return call;
+			}
+		}
+			
+		return buildBinary(e.location, e.op, lhs, rhs);
```

Comment:
I don'tt hink trying to do operator overlaoding BEFORE the normal thing makes any sense.